<!doctypehtml><html lang="en"><meta charset="utf-8"><title>Error Boundaries</title><meta name="viewport"content="width=device-width,initial-scale=1"><link rel="stylesheet"href="https://raw.githubusercontent.com/shinokada/prism-coy-theme/main/theme_common.css"><section id="error-boundaries"class="level1"><h1>Error Boundaries</h1><p>In the past, JavaScript errors inside components used to corrupt React’s internal state and cause it to <a href="https://github.com/facebook/react/issues/4026">emit</a> <a href="https://github.com/facebook/react/issues/6895">cryptic</a> <a href="https://github.com/facebook/react/issues/8579">errors</a> on next renders. These errors were always caused by an earlier error in the application code, but React did not provide a way to handle them gracefully in components, and could not recover from them.<section id="introducing-error-boundaries"class="level2"><h2>Introducing Error Boundaries</h2><p>A JavaScript error in a part of the UI shouldn’t break the whole app. To solve this problem for React users, React 16 introduces a new concept of an “error boundary”.<p>Error boundaries are React components that <strong>catch JavaScript errors anywhere in their child component tree, log those errors, and display a fallback UI</strong> instead of the component tree that crashed. Error boundaries catch errors during rendering, in lifecycle methods, and in constructors of the whole tree below them.<blockquote><p>Note<p>Error boundaries do <strong>not</strong> catch errors for:<ul><li>Event handlers (<a href="#how-about-event-handlers">learn more</a>)<li>Asynchronous code (e.g. <code>setTimeout</code> or <code>requestAnimationFrame</code> callbacks)<li>Server side rendering<li>Errors thrown in the error boundary itself (rather than its children)</ul></blockquote><p>A class component becomes an error boundary if it defines either (or both) of the lifecycle methods <a href="./react-component.html#static-getderivedstatefromerror"><code>static getDerivedStateFromError()</code></a> or <a href="./react-component.html#componentdidcatch"><code>componentDidCatch()</code></a>. Use <code>static getDerivedStateFromError()</code> to render a fallback UI after an error has been thrown. Use <code>componentDidCatch()</code> to log error information.<pre class="language-js{7-10,12-15,18-21}"><code class="language-js{7-10,12-15,18-21}">class ErrorBoundary extends React.Component {
  constructor(props) {
    super(props);
    this.state = { hasError: false };
  }

  static getDerivedStateFromError(error) {
    // Update state so the next render will show the fallback UI.
    return { hasError: true };
  }

  componentDidCatch(error, errorInfo) {
    // You can also log the error to an error reporting service
    logErrorToMyService(error, errorInfo);
  }

  render() {
    if (this.state.hasError) {
      // You can render any custom fallback UI
      return &#x3C;h1>Something went wrong.&#x3C;/h1>;
    }

    return this.props.children; 
  }
}</code></pre><p>Then you can use it as a regular component:<pre class="language-js"><code class="language-js"><span class="token operator">&#x3C;</span><span class="token maybe-class-name">ErrorBoundary</span><span class="token operator">></span>
  <span class="token operator">&#x3C;</span><span class="token maybe-class-name">MyWidget</span> <span class="token operator">/</span><span class="token operator">></span>
<span class="token operator">&#x3C;</span><span class="token operator">/</span><span class="token maybe-class-name">ErrorBoundary</span><span class="token operator">></span></code></pre><p>Error boundaries work like a JavaScript <code>catch {}</code> block, but for components. Only class components can be error boundaries. In practice, most of the time you’ll want to declare an error boundary component once and use it throughout your application.<p>Note that <strong>error boundaries only catch errors in the components below them in the tree</strong>. An error boundary can’t catch an error within itself. If an error boundary fails trying to render the error message, the error will propagate to the closest error boundary above it. This, too, is similar to how the <code>catch {}</code> block works in JavaScript.</section><section id="live-demo"class="level2"><h2>Live Demo</h2><p>Check out <a href="https://codepen.io/gaearon/pen/wqvxGa?editors=0010">this example of declaring and using an error boundary</a>.</section><section id="where-to-place-error-boundaries"class="level2"><h2>Where to Place Error Boundaries</h2><p>The granularity of error boundaries is up to you. You may wrap top-level route components to display a “Something went wrong” message to the user, just like how server-side frameworks often handle crashes. You may also wrap individual widgets in an error boundary to protect them from crashing the rest of the application.</section><section id="new-behavior-for-uncaught-errors"class="level2"><h2>New Behavior for Uncaught Errors</h2><p>This change has an important implication. <strong>As of React 16, errors that were not caught by any error boundary will result in unmounting of the whole React component tree.</strong><p>We debated this decision, but in our experience it is worse to leave corrupted UI in place than to completely remove it. For example, in a product like Messenger leaving the broken UI visible could lead to somebody sending a message to the wrong person. Similarly, it is worse for a payments app to display a wrong amount than to render nothing.<p>This change means that as you migrate to React 16, you will likely uncover existing crashes in your application that have been unnoticed before. Adding error boundaries lets you provide better user experience when something goes wrong.<p>For example, Facebook Messenger wraps content of the sidebar, the info panel, the conversation log, and the message input into separate error boundaries. If some component in one of these UI areas crashes, the rest of them remain interactive.<p>We also encourage you to use JS error reporting services (or build your own) so that you can learn about unhandled exceptions as they happen in production, and fix them.</section><section id="component-stack-traces"class="level2"><h2>Component Stack Traces</h2><p>React 16 prints all errors that occurred during rendering to the console in development, even if the application accidentally swallows them. In addition to the error message and the JavaScript stack, it also provides component stack traces. Now you can see where exactly in the component tree the failure has happened:</p><img src="./docs/error-boundaries-stack-trace.png"style="max-width:100%"alt="Error caught by Error Boundary component"><p>You can also see the filenames and line numbers in the component stack trace. This works by default in <a href="https://github.com/facebookincubator/create-react-app">Create React App</a> projects:</p><img src="./docs/error-boundaries-stack-trace-line-numbers.png"style="max-width:100%"alt="Error caught by Error Boundary component with line numbers"><p>If you don’t use Create React App, you can add <a href="https://www.npmjs.com/package/@babel/plugin-transform-react-jsx-source">this plugin</a> manually to your Babel configuration. Note that it’s intended only for development and <strong>must be disabled in production</strong>.<blockquote><p>Note<p>Component names displayed in the stack traces depend on the <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/name"><code>Function.name</code></a> property. If you support older browsers and devices which may not yet provide this natively (e.g. IE 11), consider including a <code>Function.name</code> polyfill in your bundled application, such as <a href="https://github.com/JamesMGreene/Function.name"><code>function.name-polyfill</code></a>. Alternatively, you may explicitly set the <a href="./react-component.html#displayname"><code>displayName</code></a> property on all your components.</blockquote></section><section id="how-about-trycatch"class="level2"><h2>How About try/catch?</h2><p><code>try</code> / <code>catch</code> is great but it only works for imperative code:<pre class="language-js"><code class="language-js"><span class="token keyword control-flow">try</span> <span class="token punctuation">{</span>
  <span class="token function">showButton</span><span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token punctuation">;</span>
<span class="token punctuation">}</span> <span class="token keyword control-flow">catch</span> <span class="token punctuation">(</span>error<span class="token punctuation">)</span> <span class="token punctuation">{</span>
  <span class="token comment">// ...</span>
<span class="token punctuation">}</span></code></pre><p>However, React components are declarative and specify <em>what</em> should be rendered:<pre class="language-js"><code class="language-js"><span class="token operator">&#x3C;</span><span class="token maybe-class-name">Button</span> <span class="token operator">/</span><span class="token operator">></span></code></pre><p>Error boundaries preserve the declarative nature of React, and behave as you would expect. For example, even if an error occurs in a <code>componentDidUpdate</code> method caused by a <code>setState</code> somewhere deep in the tree, it will still correctly propagate to the closest error boundary.</section><section id="how-about-event-handlers"class="level2"><h2>How About Event Handlers?</h2><p>Error boundaries <strong>do not</strong> catch errors inside event handlers.<p>React doesn't need error boundaries to recover from errors in event handlers. Unlike the render method and lifecycle methods, the event handlers don't happen during rendering. So if they throw, React still knows what to display on the screen.<p>If you need to catch an error inside an event handler, use the regular JavaScript <code>try</code> / <code>catch</code> statement:<pre class="language-js{9-13,17-20}"><code class="language-js{9-13,17-20}">class MyComponent extends React.Component {
  constructor(props) {
    super(props);
    this.state = { error: null };
    this.handleClick = this.handleClick.bind(this);
  }

  handleClick() {
    try {
      // Do something that could throw
    } catch (error) {
      this.setState({ error });
    }
  }

  render() {
    if (this.state.error) {
      return &#x3C;h1>Caught an error.&#x3C;/h1>
    }
    return &#x3C;button onClick={this.handleClick}>Click Me&#x3C;/button>
  }
}</code></pre><p>Note that the above example is demonstrating regular JavaScript behavior and doesn't use error boundaries.</section><section id="naming-changes-from-react-15"class="level2"><h2>Naming Changes from React 15</h2><p>React 15 included a very limited support for error boundaries under a different method name: <code>unstable_handleError</code>. This method no longer works, and you will need to change it to <code>componentDidCatch</code> in your code starting from the first 16 beta release.<p>For this change, we’ve provided a <a href="https://github.com/reactjs/react-codemod#error-boundaries">codemod</a> to automatically migrate your code. <span style="float:footnote"><a href="./index.html#toc">Go to TOC</a></span></section></section>